beacon-chain/rpc/over/handlers_deposit.go

```diff
@@ -0,0 +1,293 @@
+	stateId := r.PathValue("state_id")
+	if stateId == "" {
+		httputil.HandleError(w, "state_id is required in URL params", http.StatusBadRequest)
+		return
+	}
+	st, err := s.Stater.State(ctx, []byte(stateId))
+	if err != nil {
+		httputil.WriteError(w, handleWrapError(err, "could not retrieve state", http.StatusNotFound))
+		return
+	}
+
+	// Deposit estimation is only supported for Electra and later versions.
+	if st.Version() < version.Electra {
+		httputil.HandleError(w, "Deposit estimation is not supported for pre-Electra.", http.StatusBadRequest)
+		return
+	}
+
+	// Parse pubkey from URL params
+	rawPubkey := r.PathValue("pubkey")
+	if rawPubkey == "" {
+		httputil.HandleError(w, "pubkey is required in URL params", http.StatusBadRequest)
+		return
+	}
+
+	// Get metadata for response
+	isOptimistic, err := s.OptimisticModeFetcher.IsOptimistic(r.Context())
+	if err != nil {
+		httputil.WriteError(w, handleWrapError(err, "could not get optimistic mode info", http.StatusInternalServerError))
+		return
+	}
+	blockRoot, err := st.LatestBlockHeader().HashTreeRoot()
+	if err != nil {
+		httputil.WriteError(w, handleWrapError(err, "could not calculate root of latest block header", http.StatusInternalServerError))
+		return
+	}
+	isFinalized := s.FinalizationFetcher.IsFinalized(ctx, blockRoot)
+
+	// Decode pubkey
+	hexId := append0x(rawPubkey)
+	pubkey, err := hexutil.Decode(hexId)
+	if err != nil {
+		httputil.WriteError(w, handleWrapError(err, "could not decode pubkey", http.StatusInternalServerError))
+		return
+	}
+	if len(pubkey) != fieldparams.BLSPubkeyLength {
+		httputil.WriteError(w, handleWrapError(err, "invalid pubkey length", http.StatusInternalServerError))
+		return
+	}
+
+	epoch := slots.ToEpoch(st.Slot())
+	vals := st.ValidatorsReadOnly()
+	valIndex, found := st.ValidatorIndexByPubkey(bytesutil.ToBytes48(pubkey))
+
+	activeBalance := uint64(0)
+	expectedActivationEpoch := params.BeaconConfig().FarFutureEpoch
+
+	// Iterate through validators to calculate expected activation epoch and active balance.
+	for i, val := range vals {
+		valSubStatus, err := valhelpers.ValidatorSubStatus(val, epoch)
+		if err != nil {
+			httputil.WriteError(w, handleWrapError(err, "could not get validator sub status", http.StatusBadRequest))
+			return
+		}
+
+		switch valSubStatus {
+		case validator.PendingInitialized:
+			// A "fresh" validator that is added right now can be "pending_intialized" for one epoch.
+			// The activation eligibility epoch will be set to the next epoch.
+			if helpers.IsEligibleForActivationQueue(val, epoch) {
+				estimatedActivationEligibilityEpoch := epoch + 1
+				estimatedEligibleEpochForActivation := estimatedActivationEligibilityEpoch + expectedFinalityDelay
+				_expectedActivationEpoch := helpers.ActivationExitEpoch(estimatedEligibleEpochForActivation)
+				if found && primitives.ValidatorIndex(i) == valIndex {
+					expectedActivationEpoch = _expectedActivationEpoch
+				}
+			}
+		case validator.PendingQueued:
```

Comment:
Seems great! This kind of approach was needed for estimating the "pending" activation balance in the near future, but since it is assumed to use the current activation churn limit, the code can be much simpler as you suggested.